File: SampleProjects/GoT-Characters-master/Objective-C/GoT/DataSource.m
//
//  DataSource.m
//  GoT
//
//  Created by piotrom1 on 26/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

#import "DataSource.h"

@interface DataSource ()
@property(nonatomic, copy, nonnull) NSString *reuseIdentifier;
@property(nonatomic, strong, nonnull) NSMutableArray *mutableItems;
@end

@implementation DataSource

- (nonnull instancetype)
initWithCellConfigureBlock:(nullable CellConfigureBlock)configureBlock
cellReuseIdentifier:(nonnull NSString *)reuseIdentifier {
    self = [super init];
    if (self) {
        self.cellConfigureBlock = configureBlock;
        self.reuseIdentifier = reuseIdentifier;
        self.mutableItems = [[NSMutableArray alloc] init];
    }
    return self;
}

- (nonnull NSString *)cellReuseIdentifier {
    return self.reuseIdentifier;
}

- (void)addItems:(nonnull NSArray *)items {
    [self.mutableItems addObjectsFromArray:items];
    [self.mutableItems sortUsingDescriptors:@[
                                              [NSSortDescriptor sortDescriptorWithKey:@"title"
                                                                            ascending:YES]
                                              ]]; // Hack for sorting Articles by their title. Crash if anything other than
    // Article stored as item. Solution: use objc generics and protocol -
    // func keyForSorting -> String
    if (self.tableView) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.tableView reloadData];
        });
    }
}

- (void)setItems:(nonnull NSArray *)items {
    [self.mutableItems removeAllObjects];
    [self addItems:items];
}

- (nullable id)itemAtIndexPath:(nonnull NSIndexPath *)indexPath {
    id item;
    NSInteger index = indexPath.row;
    if (index < self.mutableItems.count) {
        item = self.mutableItems[index];
    }
    return item;
}

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section {
    return self.mutableItems.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell =
    [tableView dequeueReusableCellWithIdentifier:self.cellReuseIdentifier
                                    forIndexPath:indexPath];
    id item = [self itemAtIndexPath:indexPath];
    if (item) {
        self.cellConfigureBlock(cell, indexPath, item);
    }
    return cell;
}

@end

File: SampleProjects/GoT-Characters-master/Swift/GoT/DataSource.swift
//
//  DataSource.swift
//  GoT
//
//  Created by piotrom1 on 26/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

import UIKit

typealias CellConfigureBlock = (UITableViewCell, IndexPath, Any) -> Void

class DataSource: NSObject, UITableViewDataSource {
    weak var tableView: UITableView?
    var cellReuseIdentifier: String {
        return reuseIdentifier
    }
    var cellConfigureBlock: CellConfigureBlock?

    private var reuseIdentifier = ""
    private var mutableItems: [AnyHashable] = []

    init(cellConfigureBlock configureBlock: CellConfigureBlock?, cellReuseIdentifier reuseIdentifier: String) {
        super.init()
        
        cellConfigureBlock = configureBlock
        self.reuseIdentifier = reuseIdentifier
        mutableItems = [AnyHashable]()
    
    }

    func addItems(_ items: [Any]!) {
        if let anItems = items as? [AnyHashable] {
            mutableItems.append(contentsOf: anItems)
        }
        mutableItems = (mutableItems as NSArray).sortedArray(using: [NSSortDescriptor(key: "title", ascending: true)]) as? [AnyHashable] ?? mutableItems // Hack for sorting Articles by their title. Crash if anything other than
        // Article stored as item. Solution: use objc generics and protocol -
        // func keyForSorting -> String
        if tableView != nil {
            DispatchQueue.main.async(execute: {
                self.tableView?.reloadData()
            })
        }
    } // reloads tableView on main queue

    func setItems(_ items: [Any]) {
        mutableItems.removeAll()
        addItems(items)
    } // reloads tableView on main queue

    func item(at indexPath: IndexPath) -> Any? {
        var item: Any?
        let index: Int = indexPath.row
        if index < mutableItems.count {
            item = mutableItems[index]
        }
        return item
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return mutableItems.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: UITableViewCell = tableView.dequeueReusableCell(withIdentifier: cellReuseIdentifier, for: indexPath)
        let item = self.item(at: indexPath)
        if item != nil {
            cellConfigureBlock?(cell, indexPath, item!)
        }
        return cell
    }
}